rnd/autogpt_server/autogpt_server/util/lock.py

```diff
@@ -24,9 +24,5 @@ def lock(self, key: Any):
         lock.acquire()
 
     def unlock(self, key: Any):
-        with self.locks_lock:
-            if key in self.locks:
-                lock = self.locks.pop(key)
-            else:
-                return
-        lock.release()
+        if lock := self.locks.get(key):
+            lock.release()
```

Comment:
Hmm this though will cause memory leak with the ever increasing number of keys.

I think we can still do the cleanup in a safer way

Reply:
This is my proposal on fixing this, let me know what you think:

```python
import logging
from threading import Lock
from typing import Any

from expiringdict import ExpiringDict

log = logging.getLogger(__name__)


class KeyedMutex:
    """
    This class provides a mutex that can be locked and unlocked by a specific key.
    It uses an ExpiringDict to automatically clear the mutex after a specified timeout,
    in case the key is not unlocked for a specified duration, to prevent memory leaks.
    """

    def __init__(self):
        self.locks: dict[Any, tuple[Lock, int]] = ExpiringDict(
            max_len=6000, max_age_seconds=60
        )
        self.locks_lock = Lock()

    def lock(self, key: Any):
        with self.locks_lock:
            if key not in self.locks:
                self.locks[key] = (lock := Lock(), 1)
            else:
                lock, count = self.locks.pop(key)
                self.locks[key] = (lock, count + 1)

        lock.acquire()

    def unlock(self, key: Any):
        with self.locks_lock:
            if key not in self.locks:
                log.error(f">>> This should never happen! key {key} not found in locks")

            lock, count = self.locks.pop(key)
            if count > 1:
                self.locks[key] = (lock, count - 1)

        lock.release()
```

The issue was the removal of the lock from the dict ignores the waiting lock, so providing a locked counter should prevent it. Also adding a logging when such issue happen (where it should not!)

Reply:
> Hmm this though will cause memory leak with the ever increasing number of keys.

I thought that's why it uses an `ExpiringDict`?

Reply:
ExpiringDict is added to prevent the case where the lock API is used but the unlock is never used. Basically avoiding the worst case of a memory leak. The more serious problem is when the item is expiring we won't be able to release the lock (not sure if there is a removal listener for this), so we can't set a strict time-limit, and it should be used as a guardrail.

However relying on this is problematic, because It's going to be very easy for the max_length to be hit when the system is crowded, setting a very large `max_len` however wouldn't scale well, we need an eager entry removal.